data-prepper-api/src/main/java/org/opensearch/dataprepper/model/codec/InputCodec.java

```diff
@@ -16,9 +17,18 @@ public interface InputCodec {
      * Parses an {@link InputStream}. Implementors should call the {@link Consumer} for each
      * {@link Record} loaded from the {@link InputStream}.
      *
-     * @param inputStream   The input stream for the S3 object
+     * @param inputStream   The input stream for code to process
      * @param eventConsumer The consumer which handles each event from the stream
      * @throws IOException throws IOException when invalid input is received or incorrect codec name is provided
      */
     void parse(InputStream inputStream, Consumer<Record<Event>> eventConsumer) throws IOException;
+
+    /**
+     * Parses an {@link InputFile}. Implementors should call the {@link Consumer} for each
+     * {@link Record} loaded from the {@link InputFile}
+     * @param inputFile The input file for the codec to process
+     * @param eventConsumer The consumer which handles each event from the stream
+     * @throws IOException throws IOException when invalid input is received or incorrect codec name is provided
+     */
+    void parse(InputFile inputFile, Consumer<Record<Event>> eventConsumer) throws IOException;
```

Comment:
Updated to have a separate interface call InputFile and OutputFile that extends the Parquet InputFile and OutputFile interfaces.

I don't think the InputCodec should need to advertise if it requires seeking or not. In general InputStreams are not seek-able. They should be read sequentially. This is especially true of Kafka consumers. The data should be read sequentially. Files on the other hand are not required to be read sequentially. 

I like the idea of presenting either as an option. Kafka is a good example if stream being more efficient. Local files, S3, or HDFS are good example of file being more efficient. The decision to use which source should be taken in the sink or the source depending on what is more efficient.

Reply:
Thanks for making some of these changes. I'm thinking of having Parquet entirely removed from the data-prepper-api and s3-source projects.

We can have an interface in data-prepper-api which represents something similar (hopefully with only what we really need). There can be an S3 implementation similar to what you have presently. And then finally, the parquet-codecs project can adapt that interface into the Parquet `InputFile` model.

I'll see if I can put together something a little more concrete and get back to you on it.

Reply:
I agree that dependency management can be done better as we are now pulling in parquet map reduce dependencies, but I don't think that's a blocker. We can always go back and clean up the dependencies by adding our own implementations later on.

Reply:
I think a simple fix that can be applied here is:

* Copy the existing Parquet interfaces and mostly re-use them. I think there may be some opportunity to simplify as they have both `byte[]` and `ByteBuffer` fields that seem similar. This will remove Parquet from data-prepper-api which is the most important part.
* Keep the dependency within the S3 source for now. We can fix this part in a later PR.

Reply:
I don't think this is that simple. We definitely want to use the ParquetReader library because that handles a lot of situations for us. Re-implementing that library is a lot of unnecessary work. And to be able to use that library the files just need to implement InputFile. This is a pretty minimal ask. The only dependency is on Parquet Common. 

We could build a set of converters to avoid this dependency, but that adds another layer of complexity because each implementation of our InputFile needs to map to a Parquet implementation of InputFile. It's much simpler to just ensure that our implementation of an InputFile is a Parquet InputFile.

Reply:
I'm not asking to re-implement the ParquetReader. Only adapting between the two interfaces.

So, create a new interface in data-prepper-api. Then create an [adapter](https://en.wikipedia.org/wiki/Adapter_pattern) which adapts the Parquet InputFile to the data-prepper-api InputFile.

Reply:
That's a matter of opinion. It's not a blocker for adding this feature. You are welcome to make that change on top of this change.